cardano-diffusion/tests/lib/Test/Cardano/Network/Diffusion/Testnet.hs

```diff
@@ -1418,63 +1420,72 @@ prop_track_coolingToCold_demotions ioSimTracer traceNumber =
-                                                           Set.empty)
-                                  (fromJoinedOrTerminated  Set.empty
-                                                          (Set.singleton ()))
-                                  (const False)
-                                  trJoinKillSig
-
-          govInProgressDemoteToColdWhileAlive :: Signal (Maybe (Set NtNAddr))
-          govInProgressDemoteToColdWhileAlive =
-            (\isAlive inProgressDemoteToCold ->
-              if isAlive then Just inProgressDemoteToCold
-                         else Nothing
-            ) <$> trIsNodeAlive
-              <*> govInProgressDemoteToCold
-
           allInProgressDemoteToCold :: [NtNAddr]
           allInProgressDemoteToCold = Set.toList
                                     . Set.unions
-                                    . mapMaybe snd
+                                    . map snd
                                     . Signal.eventsToList
                                     . Signal.toChangeEvents
-                                    $ govInProgressDemoteToColdWhileAlive
+                                    $ govInProgressDemoteToCold
+
+
+          cmTimeWaitDoneSig :: Signal (Maybe TimeWaitPeer)
+          cmTimeWaitDoneSig =
+              Signal.fromEvents
+            . Signal.selectEvents
+                (\case (Left (CM.TrConnectionTimeWait peer))
+                         -> Just $ TimeWaitEntered (remoteAddress peer)
+                       (Left (CM.TrConnectionTimeWaitDone peer))
+                         -> Just $ TimeWaitDone (remoteAddress peer)
+                       (Right _)  -> Just ResetAll
+                       _otherwise -> Nothing)
+            . Signal.selectEvents
+                    (\case DiffusionConnectionManagerTrace tr        -> Just $ Left tr
+                           -- reset CM signal state
+                           DiffusionSimulationTrace TrKillingNode    -> Just $ Right ()
+                           _                                         -> Nothing)
+            $ events
+
+          cmTimeWaitDoneKeyedSig :: Signal (Set NtNAddr)
+          cmTimeWaitDoneKeyedSig =
+              Signal.keyedUntil
+                (\case Just (TimeWaitEntered peer) -> Set.singleton peer
+                       _otherwise                  -> Set.empty)
+                (\case Just (TimeWaitDone    peer) -> Set.singleton peer
+                       _otherwise                  -> Set.empty)
+                (\case Just ResetAll -> True
+                       _otherwise    -> False)
+                cmTimeWaitDoneSig
+
+          cmTimeWaitDoneLinger :: Signal (Set NtNAddr)
+          cmTimeWaitDoneLinger =
+            Signal.keyedLinger'
+              (\case Just (TimeWaitEntered peer) -> Just (Set.singleton peer, 30)
+                     Just ResetAll               -> Nothing
+                     _otherwise                  -> Just (Set.empty, 0))
```

Comment:
Is there a reason to arm the `cmTimeWaitDoneLinger` on `Nothing`?

`Just (TimeWaitDone peer) -> Just (Set.singleton peer, 0)` will give us better feedback if it ever fail.

Reply:
Oh, we should not return `Nothing` in the `Nothing` case, since that would reset all the lingered events.  Ignore my comment then.